File: kernel/src/arch/x86_64/pic.rs
//! Legacy 8259 PIC — we remap its vectors out of the exception range and then
//! mask all IRQs. All device interrupts we actually care about are routed
//! through the LAPIC timer (for scheduling), keeping the interrupt surface
//! tiny.

use super::port::{io_wait, outb};

const PIC1_CMD: u16 = 0x20;
const PIC1_DATA: u16 = 0x21;
const PIC2_CMD: u16 = 0xA0;
const PIC2_DATA: u16 = 0xA1;

/// Remap both PICs to vectors 0x30..0x40 (out of the CPU exception range) and
/// then mask every IRQ.
pub fn remap_and_mask() {
    unsafe {
        // ICW1: start init, cascade mode, ICW4 required.
        outb(PIC1_CMD, 0x11);
        io_wait();
        outb(PIC2_CMD, 0x11);
        io_wait();
        // ICW2: vector offsets.
        outb(PIC1_DATA, 0x30);
        io_wait();
        outb(PIC2_DATA, 0x38);
        io_wait();
        // ICW3: cascade.
        outb(PIC1_DATA, 0x04);
        io_wait();
        outb(PIC2_DATA, 0x02);
        io_wait();
        // ICW4: 8086 mode.
        outb(PIC1_DATA, 0x01);
        io_wait();
        outb(PIC2_DATA, 0x01);
        io_wait();
        // Mask all IRQs.
        outb(PIC1_DATA, 0xFF);
        outb(PIC2_DATA, 0xFF);
    }
}

File: kernel/pic.c
#include "pic.h"

/* Remap both PIC chips so IRQs 0-15 map to INT vectors 0x20-0x2F */
void pic_init(void) {
    /* Save masks */
    uint8_t m1 = inb(PIC1_DATA);
    uint8_t m2 = inb(PIC2_DATA);

    /* ICW1: start init sequence, cascade mode */
    outb(PIC1_CMD,  0x11); io_wait();
    outb(PIC2_CMD,  0x11); io_wait();
    /* ICW2: vector offsets */
    outb(PIC1_DATA, IRQ_BASE);      io_wait();   /* IRQ0-7  -> INT 32-39 */
    outb(PIC2_DATA, IRQ_BASE + 8);  io_wait();   /* IRQ8-15 -> INT 40-47 */
    /* ICW3: cascade wiring */
    outb(PIC1_DATA, 0x04); io_wait();  /* IR2 is slave */
    outb(PIC2_DATA, 0x02); io_wait();  /* slave ID 2   */
    /* ICW4: 8086 mode */
    outb(PIC1_DATA, 0x01); io_wait();
    outb(PIC2_DATA, 0x01); io_wait();

    /* Restore masks */
    outb(PIC1_DATA, m1);
    outb(PIC2_DATA, m2);
}

void pic_eoi(uint8_t irq) {
    if (irq >= 8) outb(PIC2_CMD, 0x20);
    outb(PIC1_CMD, 0x20);
}

void pic_mask(uint8_t irq) {
    uint16_t port = (irq < 8) ? PIC1_DATA : PIC2_DATA;
    uint8_t  bit  = 1 << (irq & 7);
    outb(port, inb(port) | bit);
}

void pic_unmask(uint8_t irq) {
    uint16_t port = (irq < 8) ? PIC1_DATA : PIC2_DATA;
    uint8_t  bit  = 1 << (irq & 7);
    outb(port, inb(port) & ~bit);
}
